tree.go

```diff
@@ -1266,3 +1266,155 @@ func ToDot(root VerkleNode) string {
 	root.Commit()
 	return fmt.Sprintf("digraph D {\n%s}", root.toDot("", ""))
 }
+
+type SerializedNode struct {
+	Node            VerkleNode
+	CommitmentBytes [32]byte
+	SerializedBytes []byte
+}
+
+// BatchSerialize is an optimized serialization API when multiple VerkleNodes serializations are required, and all are
+// available in memory.
+func (n *InternalNode) BatchSerialize() ([]SerializedNode, error) {
+	// Commit to the node to update all the nodes commitments.
+	n.Commit()
+
+	// Collect all nodes that we need to serialize.
+	nodes := make([]VerkleNode, 0, 1024)
+	nodes = n.collectNonHashedNodes(nodes)
+
+	// We collect all the *Point so we can batch invert the work the projective->affine transformation.
+	pointsToCompress := make([]*Point, 0, len(nodes)*NodeWidth+len(nodes))
+	pointToCompressPerNodeCount := make([]int, len(nodes))
+	for i := range nodes {
+		switch n := nodes[i].(type) {
+		case *InternalNode:
+			pointsToCompress = append(pointsToCompress, n.commitment)
+			count := 0
+			for _, c := range n.children {
+				_, isInternalNode := c.(*InternalNode)
+				_, isLeafNode := c.(*LeafNode)
+				if isInternalNode || isLeafNode {
+					pointsToCompress = append(pointsToCompress, c.Commitment())
+					count++
+				}
+			}
+			pointToCompressPerNodeCount[i] = 1 + count // Internal node commitment + all non-hashed children.
+		case *LeafNode:
+			pointsToCompress = append(pointsToCompress, n.commitment, n.c1, n.c2)
+			pointToCompressPerNodeCount[i] = 3 // Leaf node commitment + c1 + c2.
+		}
+	}
+
+	// Now we do the all transformations in a single-shot.
+	compressedPoints := banderwagon.ElementsToBytes(pointsToCompress)
```

Comment:
Now that we have in `pointsToCompress` all we need, let's do the compressed representation in a single shot _for everything_.